packages/@lwc/engine/src/framework/hot-swaps.ts

```diff
@@ -0,0 +1,226 @@
+            // by replacing the value of old template, which is used
+            // during the rendering process. If the template returned
+            // by render() is different from the previous stored template
+            // the styles will be reset, along with the content of the
+            // shadow, this way we can guarantee that the styles will be
+            // recalculated, and applied.
+            vm.cmpTemplate = () => [];
+            // forcing the vm to rehydrate in the next tick
+            markComponentAsDirty(vm);
+            scheduleRehydration(vm);
+        }
+    });
+    return true;
+}
+
+function rehydrateHotComponent(Ctor: ComponentConstructor): boolean {
+    const list = activeComponents.get(Ctor);
+    let canRefreshAllInstances = true;
+    list?.forEach((vm) => {
+        const { owner } = vm;
+        if (!isNull(owner)) {
+            if (isFalse(owner.isDirty)) {
+                // if a component class definition is swapped, we must reset
+                // the shadowRoot instance that hosts an instance of the old
+                // constructor in order to get a new element to be created based
+                // on the new constructor. this is a hacky way to force the owner's
+                // shadowRoot instance to be reset by replacing the value of old template,
+                // which is used during the rendering process. If the template returned
+                // by render() is different from the previous stored template
+                // the styles will be reset, along with the content of the
+                // shadow, this way we can guarantee that all children elements will be
+                // throw away, and new instances will be created.
+                owner.cmpTemplate = () => [];
+                // forcing the vm to rehydrate in the next tick
+                markComponentAsDirty(owner);
+                scheduleRehydration(owner);
+            }
+        } else {
+            // the hot swapping for components only work for instances of components
+            // created from a template, root elements can't be swapped because we
+            // don't have a way to force the creation of the element with the same state
+            // of the current element.
+            // Instead, we can report the problem to the caller so it can take action,
+            // for example: reload the entire page.
+            canRefreshAllInstances = false;
+        }
+    });
+    return canRefreshAllInstances;
+}
+
+export function getTemplateOrSwappedTemplate(tpl: Template): Template {
+    if (process.env.NODE_ENV === 'production') {
+        // this method should never leak to prod
+        throw new ReferenceError();
+    }
+    const visited: Set<Template> = new Set();
+    while (swappedTemplateMap.has(tpl) && !visited.has(tpl)) {
+        visited.add(tpl);
+        tpl = swappedTemplateMap.get(tpl)!;
+    }
+    return tpl;
```

Comment:
_I may be missing something here_: I am not seeing the necessity for looping while the swappedTemplateMap never changes inside the loop. Why can this not be done like this?
```suggestion
    if (swappedTemplateMap.has(tpl)) {
        tpl = swappedTemplateMap.get(tpl)!;
    }
    return tpl;
```

Reply:
nm, it is to handle transitive relationships 🤕